add support for default credential store

koshatul opened the pull request:
#### What change does this introduce?

Added a check for `credsStore` in the `${HOME}/.docker/config.json` file, at least on OSX the `auths` list is just URIs and there isn't a `credHelpers` list.

#### Why make this change?

To implement authentication when using applications that import this library on OSX and possibly other BSD/*nix OS

#### What approach will be taken?

Implement the check for a default credentials store past the registry specific `credHelpers` list.

koshatul committed:
added default credential store support and unit test

Signed-off-by: Kosh <koshatul@gmail.com>

mtrmac commented on Makefile:
```diff
@@ -63,7 +63,7 @@ clean:
 
 test:
-	@$(GPGME_ENV) GO111MODULE="on" go test $(BUILDFLAGS) -cover ./...
+	@PATH="$(PATH):$(shell pwd)/pkg/docker/config/testdata" $(GPGME_ENV) GO111MODULE="on" go test $(BUILDFLAGS) -cover ./...
```
Please don’t change global configuration for a single test like this. See e.g. how `XDG_RUNTIME_DIR` is handled in other tests, including the comments about `t.Parallel()`.

koshatul replied:
Fixed, implemented like `XDG_RUNTIME_DIR` in the unit test itself.

mtrmac commented on pkg/docker/config/config.go:
```diff
@@ -305,6 +306,11 @@ func findAuthentication(registry, path string, legacyFormat bool) (string, strin
+	if cs := auths.CredStore; cs != "" {
+		return getAuthFromCredHelper(cs, registry)
+	}
+
```
At least `SetAuthentication` and `RemoveAuthentication`, and maybe also `RemoveAllAuthentication` (not at all sure now) needs to be updated as well.

koshatul replied:
Updated, fixed the order of importance as well.

mtrmac commented on pkg/docker/config/config_test.go:
```diff
@@ -225,6 +225,73 @@ func TestGetAuth(t *testing.T) {
 	}
 }
 
+func TestGetHelperAuth(t *testing.T) {
```
This looks like something that should be fairly easy to add as one more test case in the existing `TestGetAuth`, without so much code duplication.

koshatul replied:
I had some issues with adding it, turns out whatever is the last thing there will leave its config in the "other config path", which works with the current tests because of their layout.

mtrmac commented on pkg/docker/config/testdata/helper.json:
```diff
@@ -0,0 +1,6 @@
+{
+    "auths": {
+        "foobar.example.org": {}
```
Hum, do we actually want commit to a design in which `credsStore` takes precedence over explicit entries for registries? (Note to self: What does Docker do?)

koshatul replied:
This is the (sanitised) docker config from my macbook, it seems to put a dummy entry in the auths and rely on the credstore to do it.

```json
{
  "auths" : {
    "private.registry.yoda" : {

    },
  },
  "HttpHeaders" : {
    "User-Agent" : "Docker-Client/19.03.2 (darwin)"
  },
  "stackOrchestrator" : "swarm",
  "credsStore" : "desktop"
}
```

looking at the source for docker/cli it looks like it uses the default `credStore` before falling back to the `auths` list.
https://github.com/docker/cli/blob/master/cli/config/configfile/file.go#L290

But I did have some trouble following the docker/cli code.

koshatul replied:
I changed the order of importance so it is more logical.

koshatul committed:
changed order of importance for credential stores

changed order of importance for auths.CredStore, and added functions
for using RemoveAll with credential helpers

Signed-off-by: Kosh <koshatul@gmail.com>

koshatul committed:
cleaning out all configPath before running test

Signed-off-by: Kosh <koshatul@gmail.com>

jankaluza commented:
Hi, and thank you for your contribution!

We’ve recently migrated this repository into a new **monorepo**: [containers/container-libs](https://github.com/containers/container-libs) along with other repositories

As part of this migration, this repository is no **longer accepting new Pull-Requests** and therefore this Pull-Request is being closed.

Thank you very much for your contribution. We would appreciate your continued help in migrating this PR to the new [container-libs repository](https://github.com/containers/container-libs). Please let us know if you are facing any issues.

You can read more about the migration and the reasoning behind it in our blog post: [Upcoming migration of three containers repositories to monorepo](https://blog.podman.io/2025/08/upcoming-migration-of-three-containers-repositories-to-monorepo/).

Thanks again for your work and for supporting the containers ecosystem!
